bee-storage/bee-storage-sled/src/access/fetch.rs

```
@@ -29,3 +29,14 @@ impl Fetch<MessageId, Message> for Storage {
             .map(|v| Message::unpack(&mut v.as_ref()).unwrap()))
     }
 }
+
+impl Fetch<MessageId, MessageMetadata> for Storage {
+    fn fetch(&self, message_id: &MessageId) -> Result<Option<MessageMetadata>, <Self as StorageBackend>::Error> {
+        Ok(self
+            .inner
+            .open_tree(TREE_MESSAGE_ID_TO_MESSAGE_METADATA)?
+            .get(message_id)?
+            // Unpacking from storage slice can't fail.
+            .map(|v| MessageMetadata::unpack(&mut v.as_ref()).unwrap()))
```

it can't fail as long as there's no hardware failure, right? I understand that we can use `unwrap` if we for example checked before that `option.is_some() == true`, but aren't we relying here on the storage being in a healthy state? I wonder if we should return an error here - something like a `InvalidStorageData`.

---

At this point, `.get(message_id)?` already succeeded so there is no hardware failure. What this unwrap means is that the bytes we actually fetched are not the ones we inserted, handling this would basically mean handling cosmic rays, I am not entirely sure it's necessary.

---

This could happen if the storage got corrupted somehow, right? But then again, I guess most of these unpacks could fail in that way. I'm not sure we need to handle it, it's obvious that if this ever fails there is a problem with the storage data that shouldn't be the case.

---

If such a thing was to happen, we would need to propagate the error all the way down and stop the program anyway, a panic seems easier.

---

Agreed, no point in recovering and carrying on running in that state anyway :+1:

---

The node is doing a lot of other stuff, which should be shut down in a safe manner. Instead of just crashing and burning it could end the socket connections with its peers in a friendly way, and do all kinds of other shutdown logic, before telling the user that the storage got corrupted (due to cosmic rays as usual or a broken disk). Ending the socket connections nicely gives our peers the information, that they need to look for other ones (because on our side went something wrong). If we just crash, there is still overhead on our peers trying to rescue the connection.

---

I am not 100% sur we should invest time and logic in something that may happen once in a lifetime but I'm not opposed to it. Let's see what all the others think.

---

I think for now I am fine it, because we have things more important, I guess. Maybe we should keep this as an issue though (which will then be there for years probably). I dont think hardware failure will happen so rarely, though, because many people run nodes on the cheapest possible VPS for understandably good reasons. I know I did back in the day.😊